pkg/controllers/provisioning/scheduling/scheduler.go

```diff
@@ -16,21 +16,26 @@ package scheduling
 
 import (
 	"context"
-	"fmt"
+	"time"
+
+	"go.uber.org/multierr"
+	"knative.dev/pkg/logging"
 
-	"github.com/mitchellh/hashstructure/v2"
 	"github.com/prometheus/client_golang/prometheus"
 	v1 "k8s.io/api/core/v1"
-	"knative.dev/pkg/logging"
 	"sigs.k8s.io/controller-runtime/pkg/client"
 	crmetrics "sigs.k8s.io/controller-runtime/pkg/metrics"
 
+	"github.com/aws/karpenter/pkg/utils/injection"
+
 	"github.com/aws/karpenter/pkg/apis/provisioning/v1alpha5"
 	"github.com/aws/karpenter/pkg/metrics"
-	"github.com/aws/karpenter/pkg/utils/injection"
-	"github.com/aws/karpenter/pkg/utils/resources"
 )
 
+// this is used internally only to mark why we've failed to schedule a pod to
+// provide a better user experience
+const scheduleFailureReasonKey = "karpenter.sh/scheduleFailureReason"
```

Comment:
unused?

Reply:
No, it's used in `applyTopologyConstraints()` if applying those constraints results in some unsatisfiable condition.  The scheduler then knows that the pod is not schedulable and won't bother checking for node compatibility while also reporting why it was unschedulable.